File: codes/2024/2024.09/4/8.py
# TreeNode 类型的定义
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left:TreeNode | None = None
        self.right:TreeNode | None = None

# BST的删除
def deleteNode(root:TreeNode, key:int) -> TreeNode | None:

    # 如果根节点为空，直接返回
    if root is None:
        return None

    # 如果根节点的值等于 key
    if root.val == key:

        """第一种情况：叶子结点直接删除"""
        # 如果左右孩子都为空，直接删除
        if root.left is None and root.right is None:
            return None

        """第二种情况：左孩子或者右孩子为空"""
        # 如果左孩子为空，返回右孩子
        if root.left is None:
            return root.right
        # 如果右孩子为空，返回左孩子
        if root.right is None:
            return root.left

        """走到这里说明是第三种情况：左右孩子都不为空"""
        # 如果左右孩子都不为空
        # 找到右子树的最小值
        minNode = getMin(root.right)
        # 将 root 的值替换为 minNode 的值
        root.val = minNode.val
        # 删除 minNode
        root.right = deleteNode(root.right, minNode.val)

    # 如果根节点的值小于 key
    elif root.val < key:
        if root.right is None:
            return root
        # 递归删除右子树
        root.right = deleteNode(root.right, key)
    # 如果根节点的值大于 key
    else:
        if root.left is None:
            return root
        # 递归删除左子树
        root.left = deleteNode(root.left, key)

    # 返回根节点
    return root


# 找到最小值
def getMin(node:TreeNode) -> TreeNode:
    # 找到最左边的节点，因为最左边的节点是最小的
    while node.left is not None:
        node = node.left
    return node

File: codes/2024/2024.09/1/8.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function (root) {
  const res = [];
  if (root == null) {
    return [];
  }
  traverse(root, 0, res);
  return res;
};

/**
 * @description 定义：从二叉树的根节点开始，按照「根节点、右子树、左子树」的顺序遍历二叉树
 * @param {TreeNode} root
 * @param {number} depth
 * @param {number[]} res
 */
var traverse = function (root, depth, res) {
  // 递归的终止条件，root 为空
  if (root == null) {
    return;
  }
  // 前序遍历位置
  //要点①： 先访问当前节点，再递归地访问右子树和左子树
  if (res.length == depth) {
    res.push(root.val);
  }
  // 要点②：一定要先递归访问右子树，再递归访问左子树
  traverse(root.right, depth + 1, res);
  traverse(root.left, depth + 1, res);
};
